File: src/test/java/com/uapp_llc/test/comparator/ComparatorFactory.java
package com.uapp_llc.test.comparator;

import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;

import com.uapp_llc.model.Column;
import com.uapp_llc.model.Task;

public final class ComparatorFactory {

  private static final Map<String, Comparator<?>> typeComparators = new HashMap<>();

  private ComparatorFactory() {
  }

  public static <T> Comparator<T> getComparator(Class<T> type) {
    if (type == null) {
      return null;
    }

    String typeName = type.getName();

    if (!typeComparators.containsKey(typeName)) {
      if (Column.class.equals(type)) {
        typeComparators.put(typeName, new ColumnComparator());
      }

      if (Task.class.equals(type)) {
        typeComparators.put(typeName, new TaskComparator(
            ComparatorFactory.getComparator(Column.class)
        ));
      }
    }

    return (Comparator<T>) typeComparators.get(typeName);
  }

}

File: src/test/kotlin/com/uapp_llc/test/comparator/ComparatorFactory.kt
package com.uapp_llc.test.comparator

import java.util.Comparator
import com.uapp_llc.model.Column
import com.uapp_llc.model.Task

object ComparatorFactory {

  private val typeComparators: MutableMap<String, Comparator<*>> = mutableMapOf()

  @JvmStatic
  fun <T> getComparator(type: Class<in T>): Comparator<in T?> {
    val typeName: String = type.name

    if (!typeComparators.containsKey(typeName)) {
      when (type) {
        Column::class.java -> typeComparators[typeName] = ColumnComparator()
        Task::class.java -> typeComparators[typeName] = TaskComparator(
            getComparator(Column::class.java)
        )
      }
    }

    return typeComparators[typeName] as Comparator<in T?>
  }

}
